### ISSUE TITLE
 Resummon of Felhunter, Tainted Blood & GCD?

### ISSUE THREAD

**Issue by Klimpb** [ccshiro/cc-buglist#307]
It is currently not possible to cast Spell Lock or Devour Magic instantly upon pet summon because the pet casts Tainted Blood instantly if it's on auto-cast which causes a GCD for the pet.

However as you can see in this video, he has Tainted Blood on auto-cast but he is still able to cast spell lock instantly upon summoning the Felhunter: http://webmup.com/dvYAW/

Now from the video in my last report(https://mega.co.nz/#!qZEV0YhY!q9k2nqbxlKoBUHgxPzugDjuXY4PzSVZK3k3P4iTZKFU), it does look like Tainted Blood doesn't cause a GCD

At:
00:13:08.749 - 00:13:10.087
&
00:20:20.485 - 00:20:21.583

The numbers to the left side being the frame when the new pet got summoned and the numbers to the right is when the pet successfully casted Spell Lock. Between this time, you can see that the Felhunter casts Tainted Blood by looking at the animation he's doing (red circles around him, same spell animation as amplify magic I believe).

**Comment by ccshiro** [comment]
Should probably have a delay before he auto-casts tainted blood so you have some time to use any spell you want.

**Comment by Klimpb** [comment]
@ccshiro How do you explain the felhunter casting Tainted Blood at the same time or just barely 1 second prior to casting spell lock at 00:20:20.485 - 00:20:21.583 though?

EDIT: My bad, that animation is from the warlock casting Soul Link, eh.. Feel pretty stupid now. Will try to watch the clips again and see if the pet does the tainted blood animation, which seems to be more of an attack/nod animation with some small red dots blood/fire on his head.

**Comment by ccshiro** [comment]
Looking at the video @SuddenDeath links it strikes me that the pet probably has some specific activation requirement. At ~7:30 in the video you see him cast it after being struck. Previous to that I didn't see it casted at all. Is is possible the felhunter only activates it after being hit by something?

Guess this needs some research.

**Comment by jovce17** [comment]
hmm i used to have lvl 35 lock or smthing like that on TBC eu magderidon as far i remember the felhunter wasnt spaming the tainted blood he was using it only while he was in a fight with some melee mob ..( it was attacked melee )

**Comment by nz54445** [comment]
@ccshiro 
Here you go shiro:

http://www.wowwiki.com/Tainted_Blood

This is the TBC version as it has been replaced in WOTLK:

Patch 3.0.2 (14-Oct-2008): Paranoia and Tainted Blood have been replaced with Fel Intelligence and Shadow Bite.

Direct quote:

This ability can be set to auto-cast; if so, the game's AI will automatically have the Felhunter cast Tainted Blood on itself whenever it enters into combat. It incurs a mana cost only when the Felhunter casts it; it doesn't use any more mana no matter how many hits the Felhunter takes.

This is almost definitive of how it should work and when the spell should be cast on auto-cast.

**Comment by ccshiro** [comment]
@nz54445 I read that on wowwiki too, but looking at the video provided in the OP I think I ought to have seen more tainted blood if it was when the felhunter is in combat (after all it's a short duration spell). As always, wowwiki isn't the most reliable source and I'd rather gather some occurrences of the auto-cast in videos to determine the trigger.

**Comment by nz54445** [comment]
I have a really hard time spotting the cast of tainted blood in the video. From the video the only things that i can notice is that it rarely is sent to attack (with ctrl+2) and it is rarely struck directly, but there are exceptions. Sometimes it's far away / not on the screen to actually see the animation, is there any textual indicator that it casts the spell or is it just guessing and looking at the visuals? Otherwise maybe the passive stance is interfering with it "being in combat" so to speak.

**Comment by ccshiro** [comment]
@nz54445 The visual is fairly distinct; the standard felhunter nod cast animation but with fire popping up in front of his head. That's the only indicator I was able to determine in the video.

**Comment by jovce17** [comment]
@ccshiro 
I found a Proof  that the fulhunter uses tainted blood only when he is attacked melee
for the entire video he only used it when the mob atcked him at 4:52
here is link https://youtu.be/7NagvKFLngU?t=290
also at 5:20 when he is attacked by the paladin

AT 6:30 he summons new felhunter and sends it to attack the paladin  but the new felhunter doesnt use tainted blood asap 
He uses it at 6:34 when he is attacked by the pala again
at 6:34 the attack from the pala is parried  we can conclude that he uses it only when he is attacked in general by a target in his melee range not only when he is taking dmg He is using it only when its not active too he is not spamming while he is fighting the pala

the way it should work is this 
a mob or player or what ever in the fell hunters range is targeting the fell hunter and auto attack is turned on aka will attack the fellhunter ...in that moment the felhunter uses tainted blood its not waiting for the hit to land.

Additional tainted blood uses in the video 7:30.
a wierd thing  soul link at 8:10 caused GCD on felhunters spells

**Comment by jovce17** [comment]
@ccshiro  you can notice tainted blood cast by trrigering GCD on fellhunter abilities and ofc by causing debuff on the paladin  after he attacks him

**Comment by jovce17** [comment]
another video
https://www.youtube.com/watch?v=kmM27qHgy2k
1:33 GCd causedd by activating soullink
1:36 tainted blood is being used after its attacked by the rogue

**Comment by Klimpb** [comment]
@jovce17 strange, I've never actually seen a video where the pet bar displays GCD. But I also doubt soul link causes GCD for the pet since in the video I linked the pet casts spell lock within 1 second of a soul link cast.

**Comment by jovce17** [comment]
there are 2 videos not 1. First is  from Neptulon EU, 2nd is from Bloodscalp idk if from EU or US.

And about not seen any videos ..up till now frankly i doubt that we were  gawking at the pet bar to see if does it show or not XD

**Comment by jovce17** [comment]
@SuddenDeath about your video he just re summons his pet while his pet is alive he doesnt recast soullink at all and as you can see at 0:04 the moment he uses spell lock devour magic and tainted blood turn gray for 0.5 sec as you can see  idk if its some kinda of interface bug some addon that he is using but it clearly shows that that pets GCD was activated ..devour magic and tainted blood turned gray even for a split second when he used spell lock you can see the GCD on the pets bar even in your video.
Also 0:01 in your video when he uses spell lock 1st time tainted blood goes gray for split second.
i think it clears the gray thing so fast coz he is spamming the macros.... visual interface bug .

another link http://eu.battle.net/wow/en/forum/topic/1820954268 this was in 2011 ...but i doubt that GCd on pets abilites was implemented with wotlk  it was there probably since vanilla .
It clearly says that ALL pets abilities are on GCD regardles of the size of the cd .
The only exception at the felhunter is paranoia coz its regarded as aura.

**Comment by jovce17** [comment]
@SuddenDeath  in your 2nd video the one from megaupload..
0:58 clearly shows how both devouring magic and tainted blood go on GCD when he uses spell lock
1:48 devoure magic > spell lock and tainted blood go on CD

ON 2:09 he uses tainted blood and it caused GCD on spell lock and devour magic and at 2:09-2:10 he wasnt able to use spell lock becouse of that you can clearly see him spaming the macro until it clears the GCD

**Comment by Klimpb** [comment]
@jovce17 yes, casting devour magic, spell lock & Tainted blood all cause pet GCD. Soul link however, does not.

**Comment by jovce17** [comment]
the only problem on our server is that the felhunter uses the tainted blood asap after he is summoned ....it should be reconfigured to be used only when the pet is attacked i wasmentioning times where you can see in your own video GCD on pets toolbar

**Comment by ccshiro** [comment]
Looking over the videos I think it's clear the felhunter auto-casts tainted blood when it takes damage. What's a bit more strange is that it seems the attack that caused the tainted blood to be auto-casted also triggers the proc effect.

I have implemented the first part, but I cannot cleanly implement the second without some hack that might break down in some cases. Given that the second prescribed effect has a very minimal impact, bordering completely negligible, I think this is non-issue and will consider the issue resolved now.

In other words, he now auto-casts it when taking damage, but the very same hit will not cause the proc of the not yet applied aura.

**Comment by nz54445** [comment]
While doing research the spell was mostly considered negligible in TBC due to the fact that melee never hit the felhunter, unless it was a charge or something to get to the lock, so it definitely won't impact the game too much. Nice work fixing the bug!
